File: impl/ts/src/runtime/keys.ts
/**
 * SEMP_KEYS handlers. Mirrors impl/go/internal/runtime/keys.go.
 *
 * @module
 */

import {
  KeysRequestType,
  newKeysRequest,
  newKeysResponse,
  sign as ed25519Sign,
  signSignedDoc,
  type KeysRequest,
  type KeysResponse,
  type KeysResponseResult,
  type KeyStoreRecord,
} from "@sempdev/semp/keys";
import { marshalWithElision } from "@sempdev/semp/canonical";
import type { Forwarder } from "@sempdev/semp/delivery";

import { domainOf } from "./helpers.js";
import type { SQLiteKeyStore } from "../store/keys.js";

const KeysOriginPrefix = "SEMP-KEYS-ORIGIN:";
const KeysRecordPrefix = "SEMP-KEYS-RECORD:";

export interface ClientKeysDeps {
  store: SQLiteKeyStore;
  localDomain: string;
  domainSignFP: string;
  domainSignPriv: Uint8Array;
  forwarder: Forwarder | null;
}

export interface FederationKeysDeps {
  store: SQLiteKeyStore;
  localDomain: string;
  domainSignFP: string;
  domainSignPriv: Uint8Array;
}

/** Process a SEMP_KEYS request from a client session. */
export async function handleClientKeys(
  raw: Uint8Array,
  deps: ClientKeysDeps,
): Promise<Uint8Array> {
  const req = parseKeysRequest(raw);
  const results: KeysResponseResult[] = [];
  for (const addr of req.addresses) {
    const d = domainOf(addr);
    if (d === deps.localDomain) {
      const local = lookupLocalKeys(deps.store, addr, req.include_domain_keys);
      if (local.status === "found") {
        try {
          signLocalResult(deps, local);
        } catch (err) {
          local.status = "error";
          local.error_reason =
            err instanceof Error ? err.message : String(err);
        }
      }
      results.push(local);
      continue;
    }
    // Cross-domain: forward the lookup over the cached federation
    // session via Forwarder.fetchKeys (semp-ts v0.5.1+).
    if (deps.forwarder === null) {
      results.push({
        address: addr,
        status: "not_found",
        domain: d,
        user_keys: [],
      });
      continue;
    }
    try {
      const peerReq = newKeysRequest(req.id, [addr]);
      peerReq.include_domain_keys = req.include_domain_keys;
      const peerResp = await deps.forwarder.fetchKeys(d, peerReq);
      const peerResult = peerResp.results.find((r) => r.address === addr);
      if (peerResult !== undefined) {
        results.push(peerResult);
      } else {
        results.push({
          address: addr,
          status: "not_found",
          domain: d,
          user_keys: [],
        });
      }
    } catch (err) {
      results.push({
        address: addr,
        status: "error",
        domain: d,
        user_keys: [],
        error_reason: err instanceof Error ? err.message : String(err),
      });
    }
  }
  return encodeResponse(newKeysResponse(req.id, results));
}

/** Process a SEMP_KEYS request received over a federation session. */
export function handleFederationKeys(
  raw: Uint8Array,
  deps: FederationKeysDeps,
): Uint8Array {
  const req = parseKeysRequest(raw);
  const results: KeysResponseResult[] = [];
  for (const addr of req.addresses) {
    const d = domainOf(addr);
    if (d === deps.localDomain) {
      const local = lookupLocalKeys(deps.store, addr, req.include_domain_keys);
      if (local.status === "found") {
        try {
          signLocalResult(deps, local);
        } catch (err) {
          local.status = "error";
          local.error_reason =
            err instanceof Error ? err.message : String(err);
        }
      }
      results.push(local);
      continue;
    }
    results.push({
      address: addr,
      status: "not_found",
      domain: d,
      user_keys: [],
    });
  }
  return encodeResponse(newKeysResponse(req.id, results));
}

function parseKeysRequest(raw: Uint8Array): KeysRequest {
  const obj = JSON.parse(new TextDecoder().decode(raw)) as KeysRequest;
  if (obj.type !== KeysRequestType || obj.step !== "request") {
    throw new Error(`unexpected SEMP_KEYS type/step: ${obj.type}/${obj.step}`);
  }
  if (typeof obj.include_domain_keys !== "boolean") {
    obj.include_domain_keys = true;
  }
  return obj;
}

function encodeResponse(resp: KeysResponse): Uint8Array {
  return new TextEncoder().encode(JSON.stringify(resp));
}

function lookupLocalKeys(
  store: SQLiteKeyStore,
  address: string,
  includeDomain: boolean,
): KeysResponseResult {
  const domain = domainOf(address);
  const result: KeysResponseResult = {
    address,
    status: "not_found",
    domain,
    user_keys: [],
  };
  let userKeys: KeyStoreRecord[] = [];
  try {
    userKeys = store.lookupUserKeys(address);
  } catch (err) {
    result.status = "error";
    result.error_reason = err instanceof Error ? err.message : String(err);
    return result;
  }
  if (userKeys.length === 0) {
    return result;
  }
  result.user_keys = userKeys.map((rec) => keyStoreToRecord(rec));
  result.status = "found";
  if (includeDomain) {
    const domRec = store.lookupDomainKey(domain);
    if (domRec !== null) {
      result.domain_key = keyStoreToRecord(domRec);
    }
    const encRec = store.lookupDomainEncryptionKey(domain);
    if (encRec !== null) {
      result.domain_enc_key = keyStoreToRecord(encRec);
    }
  }
  return result;
}

function keyStoreToRecord(rec: KeyStoreRecord): KeysResponseResult["user_keys"][number] {
  const out: KeysResponseResult["user_keys"][number] = {
    algorithm: rec.algorithm,
    public_key: rec.public_key,
    key_id: rec.key_id,
    key_type: rec.key_type,
  };
  if (rec.address !== undefined) {
    out.address = rec.address;
  }
  if (rec.created !== undefined) {
    out.created = rec.created;
  }
  if (rec.expires !== undefined) {
    out.expires = rec.expires;
  }
  if (rec.revocation !== undefined) {
    out.revocation = {
      reason: rec.revocation.reason,
      revoked_at: rec.revocation.revoked_at,
      ...(rec.revocation.replacement_key_id !== undefined
        ? { replacement_key_id: rec.revocation.replacement_key_id }
        : {}),
    };
  }
  return out;
}

interface SignDeps {
  localDomain: string;
  domainSignFP: string;
  domainSignPriv: Uint8Array;
}

/**
 * Apply per-record + response-level domain signatures per CLIENT.md
 * §3.3 / KEY.md §5.1.
 */
function signLocalResult(deps: SignDeps, result: KeysResponseResult): void {
  if (deps.domainSignPriv.length === 0) {
    throw new Error("runtime: no domain signing key");
  }
  // Per KEY.md §5.1, the per-record signature is computed over the
  // canonical bytes of the record with the `signatures` and
  // `revocation` fields elided, then appended to the record's
  // `signatures` array. signSignedDoc only supports object-property
  // paths so we do the elide/sign/append by hand here, mirroring
  // semp-go's keys.SignRecord.
  const signedRecords: KeysResponseResult["user_keys"] = [];
  for (const rec of result.user_keys) {
    const canonical = marshalWithElision(rec, (clone) => {
      if (
        typeof clone === "object" &&
        clone !== null &&
        !Array.isArray(clone)
      ) {
        const m = clone as Record<string, unknown>;
        delete m.signatures;
        delete m.revocation;
      }
    });
    const prefix = new TextEncoder().encode(KeysRecordPrefix);
    const signingInput = new Uint8Array(prefix.length + canonical.length);
    signingInput.set(prefix, 0);
    signingInput.set(canonical, prefix.length);
    const sig = ed25519Sign(deps.domainSignPriv, signingInput);
    const recordWithSig = JSON.parse(JSON.stringify(rec)) as Record<
      string,
      unknown
    >;
    const existingSigs = Array.isArray(recordWithSig.signatures)
      ? (recordWithSig.signatures as Record<string, unknown>[])
      : [];
    recordWithSig.signatures = [
      ...existingSigs,
      {
        signer: deps.localDomain,
        algorithm: "ed25519",
        key_id: deps.domainSignFP,
        value: Buffer.from(sig).toString("base64"),
        timestamp: isoNow(),
      },
    ];
    signedRecords.push(
      recordWithSig as unknown as KeysResponseResult["user_keys"][number],
    );
  }
  result.user_keys = signedRecords;
  // Response-level origin signature.
  const sigBlock = {
    algorithm: "ed25519",
    key_id: deps.domainSignFP,
    value: "",
  };
  const cloned: Record<string, unknown> = JSON.parse(JSON.stringify(result));
  cloned.origin_signature = sigBlock;
  const sig = signSignedDoc({
    preSignJSON: cloned,
    seed: deps.domainSignPriv,
    signaturePath: "origin_signature.value",
    prefix: KeysOriginPrefix,
  });
  const signed = sig.signedJSON as unknown as KeysResponseResult;
  if (signed.origin_signature !== undefined) {
    result.origin_signature = signed.origin_signature;
  }
}

function isoNow(): string {
  return new Date().toISOString().replace(/\.\d{3}Z$/, "Z");
}

File: impl/go/internal/runtime/keys.go
package runtime

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"time"

	"semp.dev/semp-go/keys"
	"semp.dev/semp-go/transport"
)

// domainEncKeyLookup is the optional interface a Store may implement
// to expose the domain encryption key alongside the signing key. The
// reference server's *store.SQLiteStore satisfies it.
type domainEncKeyLookup interface {
	LookupDomainEncryptionKey(domain string) *keys.Record
}

// handleClientKeys fulfills a SEMP_KEYS request from a connected
// client. Local addresses are served directly from the server's own
// store; addresses on a peer domain known to the Forwarder are
// fetched cross-domain via the federation session. Unknown domains
// return status="not_found".
//
// Reference: CLIENT.md section 5.4, section 5.4.6.
func handleClientKeys(ctx context.Context, conn transport.Conn, raw []byte, deps ClientDeps) error {
	var req keys.Request
	if err := json.Unmarshal(raw, &req); err != nil {
		return fmt.Errorf("parse SEMP_KEYS request: %w", err)
	}
	if req.Type != keys.RequestType || req.Step != keys.RequestStepRequest {
		return fmt.Errorf("unexpected SEMP_KEYS type/step: %s/%s", req.Type, req.Step)
	}

	results := make([]keys.ResponseResult, 0, len(req.Addresses))
	for _, addr := range req.Addresses {
		d := domainOf(addr)
		if d == deps.LocalDomain {
			local := lookupLocalKeys(ctx, deps.Store, addr, req.IncludeDomainKeys)
			if local.Status == keys.StatusFound {
				if err := signLocalResult(deps, &local); err != nil {
					local.Status = keys.StatusError
					local.ErrorReason = err.Error()
				}
			}
			results = append(results, local)
			continue
		}
		// Remote domain. Try the Forwarder.
		if deps.Forwarder == nil {
			results = append(results, keys.ResponseResult{
				Address: addr,
				Status:  keys.StatusNotFound,
				Domain:  d,
			})
			continue
		}
		peerReq := keys.NewRequest(newRequestID(), []string{addr})
		peerResp, err := deps.Forwarder.FetchKeys(ctx, d, peerReq)
		if err != nil {
			results = append(results, keys.ResponseResult{
				Address:     addr,
				Status:      keys.StatusError,
				Domain:      d,
				ErrorReason: err.Error(),
			})
			continue
		}
		found := false
		for _, r := range peerResp.Results {
			if r.Address == addr {
				results = append(results, r)
				found = true
				break
			}
		}
		if !found {
			results = append(results, keys.ResponseResult{
				Address: addr,
				Status:  keys.StatusNotFound,
				Domain:  d,
			})
		}
	}
	resp := keys.NewResponse(req.ID, results)
	return sendJSON(ctx, conn, resp)
}

// handleFederationKeys fulfills a SEMP_KEYS request received over a
// federation session. We only resolve addresses on our own local
// domain because federation does not multi-hop.
func handleFederationKeys(ctx context.Context, conn transport.Conn, raw []byte, deps FederationDeps) error {
	var req keys.Request
	if err := json.Unmarshal(raw, &req); err != nil {
		return fmt.Errorf("parse SEMP_KEYS request: %w", err)
	}
	if req.Type != keys.RequestType || req.Step != keys.RequestStepRequest {
		return fmt.Errorf("unexpected SEMP_KEYS type/step: %s/%s", req.Type, req.Step)
	}

	results := make([]keys.ResponseResult, 0, len(req.Addresses))
	clientLikeDeps := ClientDeps{
		Suite:          deps.Suite,
		Store:          deps.Store,
		LocalDomain:    deps.LocalDomain,
		DomainSignFP:   deps.DomainSignFP,
		DomainSignPriv: deps.DomainSignPriv,
	}
	for _, addr := range req.Addresses {
		d := domainOf(addr)
		if d == deps.LocalDomain {
			local := lookupLocalKeys(ctx, deps.Store, addr, req.IncludeDomainKeys)
			if local.Status == keys.StatusFound {
				if err := signLocalResult(clientLikeDeps, &local); err != nil {
					local.Status = keys.StatusError
					local.ErrorReason = err.Error()
				}
			}
			results = append(results, local)
			continue
		}
		// Federation does not multi-hop.
		results = append(results, keys.ResponseResult{
			Address: addr,
			Status:  keys.StatusNotFound,
			Domain:  d,
		})
	}
	resp := keys.NewResponse(req.ID, results)
	return sendJSON(ctx, conn, resp)
}

// lookupLocalKeys serves one address from the server's own keys.Store.
// Returns status="not_found" if the address has no published keys.
func lookupLocalKeys(ctx context.Context, store keys.Store, address string, includeDomain bool) keys.ResponseResult {
	domain := domainOf(address)
	result := keys.ResponseResult{
		Address: address,
		Domain:  domain,
		Status:  keys.StatusNotFound,
	}
	if store == nil {
		return result
	}
	userKeys, err := store.LookupUserKeys(ctx, address)
	if err != nil {
		result.Status = keys.StatusError
		result.ErrorReason = err.Error()
		return result
	}
	if len(userKeys) == 0 {
		return result
	}
	result.UserKeys = userKeys
	result.Status = keys.StatusFound
	if includeDomain {
		if domRec, err := store.LookupDomainKey(ctx, domain); err == nil && domRec != nil {
			result.DomainKey = domRec
		}
		if domEncLookup, ok := store.(domainEncKeyLookup); ok {
			if encRec := domEncLookup.LookupDomainEncryptionKey(domain); encRec != nil {
				result.DomainEncKey = encRec
			}
		}
	}
	return result
}

// signLocalResult applies the domain signatures required by
// CLIENT.md section 3.3 / KEY.md section 5.1 to a ResponseResult
// before returning it to a client. Two signatures are attached:
//
//  1. A per-record domain signature on every user Record in
//     result.UserKeys, via keys.SignRecord.
//  2. A response-level OriginSignature on the whole result via
//     keys.SignResponseResult.
//
// Records are deep-copied before signing so we never mutate the
// store's shared copies.
func signLocalResult(deps ClientDeps, result *keys.ResponseResult) error {
	if deps.DomainSignPriv == nil {
		return errors.New("runtime: no domain signing key")
	}
	signer := deps.Suite.Signer()
	cloned := make([]*keys.Record, 0, len(result.UserKeys))
	for _, rec := range result.UserKeys {
		if rec == nil {
			continue
		}
		cp := *rec
		cp.Signatures = nil
		if err := keys.SignRecord(signer, deps.DomainSignPriv, deps.LocalDomain, deps.DomainSignFP, &cp); err != nil {
			return fmt.Errorf("sign user record %s: %w", rec.KeyID, err)
		}
		cloned = append(cloned, &cp)
	}
	result.UserKeys = cloned
	if err := keys.SignResponseResult(signer, deps.DomainSignPriv, deps.DomainSignFP, result); err != nil {
		return fmt.Errorf("sign response result: %w", err)
	}
	return nil
}

// newRequestID returns a short pseudo-ULID for a SEMP_KEYS request.
// Per-session uniqueness is sufficient for correlation; we use a
// timestamp + the address count to keep the function dependency-free.
func newRequestID() string {
	return fmt.Sprintf("req-%d", time.Now().UnixNano())
}
